{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9b46170d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "from datetime import datetime, timedelta\n",
    "from dateutil.relativedelta import *\n",
    "import pandas as pd\n",
    "import os\n",
    "import csv\n",
    "from scipy.stats import pearsonr\n",
    "import numpy as np\n",
    "import re\n",
    "import pytrends\n",
    "import bash\n",
    "import random\n",
    "import pdb\n",
    "import calendar\n",
    "import requests\n",
    "from requests.exceptions import Timeout\n",
    "from pytrends.exceptions import ResponseError\n",
    "from torpy.circuit import  CellTimeoutError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "21a1afcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cookie():\n",
    "    from selenium import webdriver\n",
    "    from selenium.webdriver.edge.service import Service\n",
    "    options = webdriver.EdgeOptions()\n",
    "    service = Service(executable_path='/Users/liuy5/Documents/GitHub/GT-MVT-Annually-DMA/msedgedriver')\n",
    "    \n",
    "\n",
    "# Create a WebDriver instance for Microsoft Edge\n",
    "    driver = webdriver.Edge(service=service, options=options )\n",
    "    driver.get(\"https://trends.google.com/\")\n",
    "    time.sleep(5)\n",
    "    cookie = driver.get_cookie(\"NID\")[\"value\"]\n",
    "    driver.quit()\n",
    "    return cookie\n",
    "\n",
    "nid_cookie = f\"NID={get_cookie()}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f5a667db",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pytrends.request import TrendReq\n",
    "\n",
    "requests_args = {\n",
    "    'headers' : {\"User-Agent\":\"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36\",\n",
    "        \"Cookie\": f\"{nid_cookie}\",\n",
    "    }\n",
    "}\n",
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "26b226d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def init():\n",
    "    from torpy import TorClient\n",
    "\n",
    "    hostname = 'ifconfig.me'  # It's possible use onion hostname here as well\n",
    "    with TorClient() as tor:\n",
    "        # Choose random guard node and create 3-hops circuit\n",
    "        with tor.create_circuit(3) as circuit:\n",
    "            # Create tor stream to host\n",
    "            with circuit.create_stream((hostname, 80)) as stream:\n",
    "                # Now we can communicate with host\n",
    "                stream.send(b'GET / HTTP/1.0\\r\\nHost: %s\\r\\n\\r\\n' % hostname.encode())\n",
    "                recv = stream.recv(1024)\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "eecf1545",
   "metadata": {},
   "outputs": [],
   "source": [
    "def column_name_to_mvt_and_remove_zero_values(df, year):\n",
    "    df.rename(columns={df.columns[0]: f\"MVT_GT_{year}\" }, inplace = True)\n",
    "    df = df[df[f\"MVT_GT_{year}\"] != 0]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a432a8ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gt_by_region_by_year(mvt_keyword, start_year, end_year, geo_level):\n",
    "    df = pd.DataFrame()\n",
    "    for i in range(start_year, end_year+1):\n",
    "            try:\n",
    "                pytrend.build_payload(mvt_keyword, cat=0, timeframe=f\"{i}-01-01 {i}-12-31\", geo = \"US\")\n",
    "                temp_df = pytrend.interest_by_region(resolution=geo_level, inc_low_vol=True, inc_geo_code=False)\n",
    "                print(f\"{time.now().strf()} Obtaining GT MVT from {i}-01-01 to {i}-12-31 in {geo_level}\")\n",
    "                \n",
    "            except (ResponseError, Timeout, ConnectionAbortedError, ConnectionError) as error:\n",
    "                try:\n",
    "                    print(f\"GT Error Occured: {error}\")\n",
    "                    !pip install pytrends --upgrade\n",
    "                    #init()\n",
    "                    #rint(f\"Let's Go, Onion!\")\n",
    "                    pytrend.build_payload(mvt_keyword, cat=0, timeframe=f\"{i}-01-01 {i}-12-31\", geo = \"US\")\n",
    "                    temp_df = pytrend.interest_by_region(resolution=geo_level, inc_low_vol=True, inc_geo_code=False)\n",
    "                    print(f\"Obtaining GT MVT from {i}-01-01 to {i}-12-31 in {geo_level}\")\n",
    "\n",
    "                except (ResponseError, Timeout, ConnectionAbortedError, ConnectionError) as error:\n",
    "                    print(f\"GT Error Occured: {error}\")\n",
    "                    print(f\"Let me sleep for 1 day, till {(datetime.now() + timedelta(days=1)).strftime('%m/%d/%Y, %H:%M:%S')}\")\n",
    "                    print(\"ZZzzzz...\")\n",
    "                    time.sleep(random.randint(60*60*24, 60*60*24+30,))\n",
    "                    print(\"Was a nice sleep, now let me continue...\")\n",
    "\n",
    "                    pytrend.build_payload(mvt_keyword, cat=0, timeframe=f\"{i}-01-01 {i}-12-31\", geo = \"US\")\n",
    "                    temp_df = pytrend.interest_by_region(resolution=geo_level, inc_low_vol=True, inc_geo_code=False)\n",
    "                    print(f\"Obtaining GT MVT from {i}-01-01 to {i}-12-31 in {geo_level}\")\n",
    "                \n",
    "            temp_df = column_name_to_mvt_and_remove_zero_values(temp_df, i)\n",
    "            df = pd.concat([df, temp_df.iloc[:, 0]], axis = 1)\n",
    "            time.sleep(random.randint(600, 660))\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a5a6a136",
   "metadata": {},
   "outputs": [],
   "source": [
    "def execute_init():\n",
    "    while True:\n",
    "        try:\n",
    "            init()\n",
    "            print(\"init status completed\")\n",
    "            break\n",
    "        except (TimeoutError, CellTimeoutError, ConnectionResetError) as error:\n",
    "            print(f\"Init Error Occured: {error}\")\n",
    "            print(\"Let me sleep for 10 seconds\")\n",
    "            print(\"ZZzzzz...\")\n",
    "            time.sleep(10)\n",
    "            print(\"Was a nice sleep, now let me continue...\")\n",
    "            pass\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bd4bb7b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def GT_temp_df(start_date, end_date, geo_code, geo_name):\n",
    "    temp_df = pd.DataFrame()\n",
    "    while len(temp_df) == 0:\n",
    "        try:\n",
    "            pytrend.build_payload(mvt_keyword, cat=0, timeframe=f\"{start_date} {end_date}\", geo = \"US\")\n",
    "            temp_df = pytrend.interest_by_region(resolution='REGION', inc_low_vol=True, inc_geo_code=False)\n",
    "\n",
    "        except (ResponseError, Timeout, ConnectionAbortedError, ConnectionError) as error:\n",
    "            try:\n",
    "                print(f\"GT Error Occured: {error}\")\n",
    "                !pip install pytrends --upgrade\n",
    "                #init()\n",
    "                #rint(f\"Let's Go, Onion!\")\n",
    "                pytrend.build_payload(mvt_keyword, cat=0, timeframe=f\"{start_date} {end_date}\", geo = \"US\")\n",
    "                temp_df = pytrend.interest_by_region(resolution='REGION', inc_low_vol=True, inc_geo_code=False)\n",
    "\n",
    "            except (ResponseError, Timeout, ConnectionAbortedError, ConnectionError) as error:\n",
    "                print(f\"GT Error Occured: {error}\")\n",
    "                print(f\"Let me sleep for 1 day, till {(datetime.now() + timedelta(days=1)).strftime('%m/%d/%Y, %H:%M:%S')}\")\n",
    "                print(\"ZZzzzz...\")\n",
    "                time.sleep(random.randint(60*60*24, 60*60*24+30,))\n",
    "                print(\"Was a nice sleep, now let me continue...\")\n",
    "\n",
    "\n",
    "    print(f\"Time {pd.Timestamp.now().strftime('%Y%m%d_%H_%M')} Obtaining GT MVT from {start_date} to {end_date} in {geo_name} metro area\")\n",
    "    time.sleep(random.randint(600, 660))\n",
    "    temp_df = column_name_to_mvt_and_remove_zero_values(temp_df, geo_name)\n",
    "    return temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7575af2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_df_and_temp_df_and_rescale_by_overlap_factors(df, temp_df):\n",
    "    df = pd.concat([df, temp_df.iloc[:, 0]], axis = 1)\n",
    "    if len(df.columns) > 1:\n",
    "        factor_data = df.dropna(how = \"any\")\n",
    "        factor_divided = factor_data.iloc[:, 0]/factor_data.iloc[:, 1]\n",
    "        adjusted_factor = factor_divided.mean()\n",
    "        df.iloc[:,1] = df.iloc[:,1]*adjusted_factor\n",
    "        df = df.mean(axis = 1)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "66539e2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Errno 61] Connection refused\n",
      "ERROR:root:[ignored]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 63, in connect\n",
      "    self._socket.connect((self._router.ip, self._router.or_port))\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1412, in connect\n",
      "    self._real_connect(addr, False)\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1399, in _real_connect\n",
      "    super().connect(addr)\n",
      "ConnectionRefusedError: [Errno 61] Connection refused\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/utils.py\", line 79, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 235, in renew\n",
      "    raw_string = self.download_consensus(prev_hash)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 183, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 390, in download_consensus\n",
      "    with self._get_dir_client() as dir_client:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 375, in _get_dir_client\n",
      "    self._dir_guard, self._dir_circuit = self._create_dir_circuit(purpose='Internal dir client')\n",
      "                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 365, in _create_dir_circuit\n",
      "    guard = TorGuard(router, purpose=purpose)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/guard.py\", line 66, in __init__\n",
      "    self.__tor_socket.connect()\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 69, in connect\n",
      "    raise TorSocketConnectError(e)\n",
      "torpy.cell_socket.TorSocketConnectError: [Errno 61] Connection refused\n",
      "WARNING:torpy.utils:Retry with another router...\n",
      "ERROR:torpy.cell_socket:timed out\n",
      "ERROR:root:[ignored]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 63, in connect\n",
      "    self._socket.connect((self._router.ip, self._router.or_port))\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1412, in connect\n",
      "    self._real_connect(addr, False)\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1399, in _real_connect\n",
      "    super().connect(addr)\n",
      "TimeoutError: timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/utils.py\", line 79, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 235, in renew\n",
      "    raw_string = self.download_consensus(prev_hash)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 183, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 390, in download_consensus\n",
      "    with self._get_dir_client() as dir_client:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 375, in _get_dir_client\n",
      "    self._dir_guard, self._dir_circuit = self._create_dir_circuit(purpose='Internal dir client')\n",
      "                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 365, in _create_dir_circuit\n",
      "    guard = TorGuard(router, purpose=purpose)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/guard.py\", line 66, in __init__\n",
      "    self.__tor_socket.connect()\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 69, in connect\n",
      "    raise TorSocketConnectError(e)\n",
      "torpy.cell_socket.TorSocketConnectError: timed out\n",
      "WARNING:torpy.utils:Retry with another router...\n",
      "ERROR:torpy.cell_socket:timed out\n",
      "ERROR:root:[ignored]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 63, in connect\n",
      "    self._socket.connect((self._router.ip, self._router.or_port))\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1412, in connect\n",
      "    self._real_connect(addr, False)\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1399, in _real_connect\n",
      "    super().connect(addr)\n",
      "TimeoutError: timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/utils.py\", line 79, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 235, in renew\n",
      "    raw_string = self.download_consensus(prev_hash)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 183, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 390, in download_consensus\n",
      "    with self._get_dir_client() as dir_client:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 375, in _get_dir_client\n",
      "    self._dir_guard, self._dir_circuit = self._create_dir_circuit(purpose='Internal dir client')\n",
      "                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 365, in _create_dir_circuit\n",
      "    guard = TorGuard(router, purpose=purpose)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/guard.py\", line 66, in __init__\n",
      "    self.__tor_socket.connect()\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 69, in connect\n",
      "    raise TorSocketConnectError(e)\n",
      "torpy.cell_socket.TorSocketConnectError: timed out\n",
      "WARNING:torpy.utils:Retry with another router...\n",
      "ERROR:torpy.cell_socket:timed out\n",
      "ERROR:root:[ignored]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 63, in connect\n",
      "    self._socket.connect((self._router.ip, self._router.or_port))\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1412, in connect\n",
      "    self._real_connect(addr, False)\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1399, in _real_connect\n",
      "    super().connect(addr)\n",
      "TimeoutError: timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/utils.py\", line 79, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 235, in renew\n",
      "    raw_string = self.download_consensus(prev_hash)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 183, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 390, in download_consensus\n",
      "    with self._get_dir_client() as dir_client:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 375, in _get_dir_client\n",
      "    self._dir_guard, self._dir_circuit = self._create_dir_circuit(purpose='Internal dir client')\n",
      "                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 365, in _create_dir_circuit\n",
      "    guard = TorGuard(router, purpose=purpose)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/guard.py\", line 66, in __init__\n",
      "    self.__tor_socket.connect()\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 69, in connect\n",
      "    raise TorSocketConnectError(e)\n",
      "torpy.cell_socket.TorSocketConnectError: timed out\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:torpy.utils:Retry with another router...\n",
      "ERROR:torpy.cell_socket:[Errno 61] Connection refused\n",
      "ERROR:root:[ignored]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 63, in connect\n",
      "    self._socket.connect((self._router.ip, self._router.or_port))\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1412, in connect\n",
      "    self._real_connect(addr, False)\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1399, in _real_connect\n",
      "    super().connect(addr)\n",
      "ConnectionRefusedError: [Errno 61] Connection refused\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/utils.py\", line 79, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 235, in renew\n",
      "    raw_string = self.download_consensus(prev_hash)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 183, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 390, in download_consensus\n",
      "    with self._get_dir_client() as dir_client:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 375, in _get_dir_client\n",
      "    self._dir_guard, self._dir_circuit = self._create_dir_circuit(purpose='Internal dir client')\n",
      "                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 365, in _create_dir_circuit\n",
      "    guard = TorGuard(router, purpose=purpose)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/guard.py\", line 66, in __init__\n",
      "    self.__tor_socket.connect()\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 69, in connect\n",
      "    raise TorSocketConnectError(e)\n",
      "torpy.cell_socket.TorSocketConnectError: [Errno 61] Connection refused\n",
      "WARNING:torpy.utils:Retry with another router...\n",
      "ERROR:torpy.utils:[ignored] torpy.cell_socket.TorSocketConnectError: [Errno 61] Connection refused\n",
      "WARNING:torpy.utils:Retry with another guard...\n",
      "ERROR:torpy.cell_socket:timed out\n",
      "ERROR:root:[ignored]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 63, in connect\n",
      "    self._socket.connect((self._router.ip, self._router.or_port))\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1412, in connect\n",
      "    self._real_connect(addr, False)\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1399, in _real_connect\n",
      "    super().connect(addr)\n",
      "TimeoutError: timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/utils.py\", line 79, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 235, in renew\n",
      "    raw_string = self.download_consensus(prev_hash)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 183, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 390, in download_consensus\n",
      "    with self._get_dir_client() as dir_client:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 375, in _get_dir_client\n",
      "    self._dir_guard, self._dir_circuit = self._create_dir_circuit(purpose='Internal dir client')\n",
      "                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 365, in _create_dir_circuit\n",
      "    guard = TorGuard(router, purpose=purpose)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/guard.py\", line 66, in __init__\n",
      "    self.__tor_socket.connect()\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 69, in connect\n",
      "    raise TorSocketConnectError(e)\n",
      "torpy.cell_socket.TorSocketConnectError: timed out\n",
      "WARNING:torpy.utils:Retry with another router...\n",
      "ERROR:torpy.cell_socket:timed out\n",
      "ERROR:root:[ignored]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 63, in connect\n",
      "    self._socket.connect((self._router.ip, self._router.or_port))\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1412, in connect\n",
      "    self._real_connect(addr, False)\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/ssl.py\", line 1399, in _real_connect\n",
      "    super().connect(addr)\n",
      "TimeoutError: timed out\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/utils.py\", line 79, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 235, in renew\n",
      "    raw_string = self.download_consensus(prev_hash)\n",
      "                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 183, in newfn\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 390, in download_consensus\n",
      "    with self._get_dir_client() as dir_client:\n",
      "         ^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 375, in _get_dir_client\n",
      "    self._dir_guard, self._dir_circuit = self._create_dir_circuit(purpose='Internal dir client')\n",
      "                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/consesus.py\", line 365, in _create_dir_circuit\n",
      "    guard = TorGuard(router, purpose=purpose)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/guard.py\", line 66, in __init__\n",
      "    self.__tor_socket.connect()\n",
      "  File \"/Users/liuy5/anaconda3/lib/python3.11/site-packages/torpy/cell_socket.py\", line 69, in connect\n",
      "    raise TorSocketConnectError(e)\n",
      "torpy.cell_socket.TorSocketConnectError: timed out\n",
      "WARNING:torpy.utils:Retry with another router...\n",
      "WARNING:torpy.consesus:Unknown voter present\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "init()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0f3e3a90",
   "metadata": {},
   "outputs": [],
   "source": [
    "#basic pytrend settings\n",
    "pytrend = TrendReq(retries=3, requests_args = requests_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7e8795a5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nkw_list = [\"car stolen+find stolen car+report police stolen car+insurance car stolen-dream-check\",\\n    \"I raped+raped me+raped report police+being raped+been raped - dream - kobe - trump - porn\",\\n    \"my stolen+find my stolen+theft report police+someone stolen my - car - heart - dream - number\",\\n    \"burglary+home burglarized+my house broken into+my home broken into+burglary report police - dream\",\\n    \"my committed a crime+police arrested my+my arrested+my in jail+my in prison - arrested development\",\\n    \"I hate family+parents hate me+I hate my parents+I hate school+I hate teacher+I hate dad+I hate mom\"]\\n'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "kw_list = [\"car stolen+find stolen car+report police stolen car+insurance car stolen-dream-check\",\n",
    "    \"I raped+raped me+raped report police+being raped+been raped - dream - kobe - trump - porn\",\n",
    "    \"my stolen+find my stolen+theft report police+someone stolen my - car - heart - dream - number\",\n",
    "    \"burglary+home burglarized+my house broken into+my home broken into+burglary report police - dream\",\n",
    "    \"my committed a crime+police arrested my+my arrested+my in jail+my in prison - arrested development\",\n",
    "    \"I hate family+parents hate me+I hate my parents+I hate school+I hate teacher+I hate dad+I hate mom\"]\n",
    "'''\n",
    "#years = [i for i in range(2010, 2016)]\n",
    "#print(years)\n",
    "#months = [i for i in range(1, 13, 3)] \n",
    "#print(months)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30921c43",
   "metadata": {},
   "source": [
    "### GT Keywords (Test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2105a9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mvt_keyword = [\"car stolen+find stolen car+report police stolen car+insurance car stolen-dream-check\"] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "753dfe1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gt_by_region_by_year(mvt_keyword, 2011, 2014, \"REGION\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "720ed6be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reversed_dict = {value: key for key, value in region_code_dict.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "165194c8",
   "metadata": {},
   "source": [
    "### Dates (From, To)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5a40f85c",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_year = 2011\n",
    "end_year = 2022\n",
    "geo_level = \"DMA\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6f40044",
   "metadata": {},
   "source": [
    "## Execute Pytrends to Pull Daily Data from Google Trends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "48a26924",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Obtaining GT MVT from 2011-01-01 to 2011-12-31 in DMA\n",
      "Obtaining GT MVT from 2012-01-01 to 2012-12-31 in DMA\n",
      "Obtaining GT MVT from 2013-01-01 to 2013-12-31 in DMA\n",
      "Obtaining GT MVT from 2014-01-01 to 2014-12-31 in DMA\n",
      "Obtaining GT MVT from 2015-01-01 to 2015-12-31 in DMA\n",
      "Obtaining GT MVT from 2016-01-01 to 2016-12-31 in DMA\n",
      "Obtaining GT MVT from 2017-01-01 to 2017-12-31 in DMA\n",
      "Obtaining GT MVT from 2018-01-01 to 2018-12-31 in DMA\n",
      "Obtaining GT MVT from 2019-01-01 to 2019-12-31 in DMA\n",
      "Obtaining GT MVT from 2020-01-01 to 2020-12-31 in DMA\n",
      "Obtaining GT MVT from 2021-01-01 to 2021-12-31 in DMA\n",
      "Obtaining GT MVT from 2022-01-01 to 2022-12-31 in DMA\n",
      "DMA_2011_2022_annually_20240226_20_00 file saved\n",
      "Obtaining GT MVT from 2011-01-01 to 2011-12-31 in DMA\n",
      "Obtaining GT MVT from 2012-01-01 to 2012-12-31 in DMA\n",
      "Obtaining GT MVT from 2013-01-01 to 2013-12-31 in DMA\n",
      "Obtaining GT MVT from 2014-01-01 to 2014-12-31 in DMA\n",
      "Obtaining GT MVT from 2015-01-01 to 2015-12-31 in DMA\n",
      "Obtaining GT MVT from 2016-01-01 to 2016-12-31 in DMA\n",
      "Obtaining GT MVT from 2017-01-01 to 2017-12-31 in DMA\n",
      "Obtaining GT MVT from 2018-01-01 to 2018-12-31 in DMA\n",
      "Obtaining GT MVT from 2019-01-01 to 2019-12-31 in DMA\n",
      "Obtaining GT MVT from 2020-01-01 to 2020-12-31 in DMA\n",
      "Obtaining GT MVT from 2021-01-01 to 2021-12-31 in DMA\n",
      "Obtaining GT MVT from 2022-01-01 to 2022-12-31 in DMA\n",
      "DMA_2011_2022_annually_20240227_01_00 file saved\n",
      "Obtaining GT MVT from 2011-01-01 to 2011-12-31 in DMA\n",
      "Obtaining GT MVT from 2012-01-01 to 2012-12-31 in DMA\n",
      "Obtaining GT MVT from 2013-01-01 to 2013-12-31 in DMA\n",
      "Obtaining GT MVT from 2014-01-01 to 2014-12-31 in DMA\n",
      "Obtaining GT MVT from 2015-01-01 to 2015-12-31 in DMA\n",
      "Obtaining GT MVT from 2016-01-01 to 2016-12-31 in DMA\n",
      "Obtaining GT MVT from 2017-01-01 to 2017-12-31 in DMA\n",
      "Obtaining GT MVT from 2018-01-01 to 2018-12-31 in DMA\n",
      "Obtaining GT MVT from 2019-01-01 to 2019-12-31 in DMA\n",
      "Obtaining GT MVT from 2020-01-01 to 2020-12-31 in DMA\n",
      "Obtaining GT MVT from 2021-01-01 to 2021-12-31 in DMA\n",
      "Obtaining GT MVT from 2022-01-01 to 2022-12-31 in DMA\n",
      "DMA_2011_2022_annually_20240227_05_58 file saved\n"
     ]
    },
    {
     "ename": "RetryError",
     "evalue": "HTTPSConnectionPool(host='trends.google.com', port=443): Max retries exceeded with url: /trends/api/widgetdata/comparedgeo?req=%7B%22geo%22%3A+%7B%22country%22%3A+%22US%22%7D%2C+%22comparisonItem%22%3A+%5B%7B%22time%22%3A+%222011-01-01+2011-12-31%22%2C+%22complexKeywordsRestriction%22%3A+%7B%22keyword%22%3A+%5B%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22car+stolen%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22find+stolen+car%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22report+police+stolen+car%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22insurance+car+stolen-dream-check%22%7D%5D%2C+%22operator%22%3A+%22OR%22%7D%7D%5D%2C+%22resolution%22%3A+%22DMA%22%2C+%22locale%22%3A+%22en-US%22%2C+%22requestOptions%22%3A+%7B%22property%22%3A+%22%22%2C+%22backend%22%3A+%22IZG%22%2C+%22category%22%3A+0%7D%2C+%22userConfig%22%3A+%7B%22userType%22%3A+%22USER_TYPE_SCRAPER%22%7D%2C+%22includeLowSearchVolumeGeos%22%3A+true%7D&token=APP6_UEAAAAAZd86w2fv-SZP28rOlNlhEpDIhsrwrJX-&tz=360 (Caused by ResponseError('too many 429 error responses'))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/requests/adapters.py:486\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    485\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 486\u001b[0m     resp \u001b[38;5;241m=\u001b[39m conn\u001b[38;5;241m.\u001b[39murlopen(\n\u001b[1;32m    487\u001b[0m         method\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mmethod,\n\u001b[1;32m    488\u001b[0m         url\u001b[38;5;241m=\u001b[39murl,\n\u001b[1;32m    489\u001b[0m         body\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mbody,\n\u001b[1;32m    490\u001b[0m         headers\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mheaders,\n\u001b[1;32m    491\u001b[0m         redirect\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    492\u001b[0m         assert_same_host\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    493\u001b[0m         preload_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    494\u001b[0m         decode_content\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    495\u001b[0m         retries\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmax_retries,\n\u001b[1;32m    496\u001b[0m         timeout\u001b[38;5;241m=\u001b[39mtimeout,\n\u001b[1;32m    497\u001b[0m         chunked\u001b[38;5;241m=\u001b[39mchunked,\n\u001b[1;32m    498\u001b[0m     )\n\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/urllib3/connectionpool.py:889\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    888\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRetry: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, url)\n\u001b[0;32m--> 889\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39murlopen(\n\u001b[1;32m    890\u001b[0m         method,\n\u001b[1;32m    891\u001b[0m         url,\n\u001b[1;32m    892\u001b[0m         body,\n\u001b[1;32m    893\u001b[0m         headers,\n\u001b[1;32m    894\u001b[0m         retries\u001b[38;5;241m=\u001b[39mretries,\n\u001b[1;32m    895\u001b[0m         redirect\u001b[38;5;241m=\u001b[39mredirect,\n\u001b[1;32m    896\u001b[0m         assert_same_host\u001b[38;5;241m=\u001b[39massert_same_host,\n\u001b[1;32m    897\u001b[0m         timeout\u001b[38;5;241m=\u001b[39mtimeout,\n\u001b[1;32m    898\u001b[0m         pool_timeout\u001b[38;5;241m=\u001b[39mpool_timeout,\n\u001b[1;32m    899\u001b[0m         release_conn\u001b[38;5;241m=\u001b[39mrelease_conn,\n\u001b[1;32m    900\u001b[0m         chunked\u001b[38;5;241m=\u001b[39mchunked,\n\u001b[1;32m    901\u001b[0m         body_pos\u001b[38;5;241m=\u001b[39mbody_pos,\n\u001b[1;32m    902\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mresponse_kw\n\u001b[1;32m    903\u001b[0m     )\n\u001b[1;32m    905\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/urllib3/connectionpool.py:889\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    888\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRetry: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, url)\n\u001b[0;32m--> 889\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39murlopen(\n\u001b[1;32m    890\u001b[0m         method,\n\u001b[1;32m    891\u001b[0m         url,\n\u001b[1;32m    892\u001b[0m         body,\n\u001b[1;32m    893\u001b[0m         headers,\n\u001b[1;32m    894\u001b[0m         retries\u001b[38;5;241m=\u001b[39mretries,\n\u001b[1;32m    895\u001b[0m         redirect\u001b[38;5;241m=\u001b[39mredirect,\n\u001b[1;32m    896\u001b[0m         assert_same_host\u001b[38;5;241m=\u001b[39massert_same_host,\n\u001b[1;32m    897\u001b[0m         timeout\u001b[38;5;241m=\u001b[39mtimeout,\n\u001b[1;32m    898\u001b[0m         pool_timeout\u001b[38;5;241m=\u001b[39mpool_timeout,\n\u001b[1;32m    899\u001b[0m         release_conn\u001b[38;5;241m=\u001b[39mrelease_conn,\n\u001b[1;32m    900\u001b[0m         chunked\u001b[38;5;241m=\u001b[39mchunked,\n\u001b[1;32m    901\u001b[0m         body_pos\u001b[38;5;241m=\u001b[39mbody_pos,\n\u001b[1;32m    902\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mresponse_kw\n\u001b[1;32m    903\u001b[0m     )\n\u001b[1;32m    905\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/urllib3/connectionpool.py:889\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    888\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRetry: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, url)\n\u001b[0;32m--> 889\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39murlopen(\n\u001b[1;32m    890\u001b[0m         method,\n\u001b[1;32m    891\u001b[0m         url,\n\u001b[1;32m    892\u001b[0m         body,\n\u001b[1;32m    893\u001b[0m         headers,\n\u001b[1;32m    894\u001b[0m         retries\u001b[38;5;241m=\u001b[39mretries,\n\u001b[1;32m    895\u001b[0m         redirect\u001b[38;5;241m=\u001b[39mredirect,\n\u001b[1;32m    896\u001b[0m         assert_same_host\u001b[38;5;241m=\u001b[39massert_same_host,\n\u001b[1;32m    897\u001b[0m         timeout\u001b[38;5;241m=\u001b[39mtimeout,\n\u001b[1;32m    898\u001b[0m         pool_timeout\u001b[38;5;241m=\u001b[39mpool_timeout,\n\u001b[1;32m    899\u001b[0m         release_conn\u001b[38;5;241m=\u001b[39mrelease_conn,\n\u001b[1;32m    900\u001b[0m         chunked\u001b[38;5;241m=\u001b[39mchunked,\n\u001b[1;32m    901\u001b[0m         body_pos\u001b[38;5;241m=\u001b[39mbody_pos,\n\u001b[1;32m    902\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mresponse_kw\n\u001b[1;32m    903\u001b[0m     )\n\u001b[1;32m    905\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/urllib3/connectionpool.py:879\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    878\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 879\u001b[0m     retries \u001b[38;5;241m=\u001b[39m retries\u001b[38;5;241m.\u001b[39mincrement(method, url, response\u001b[38;5;241m=\u001b[39mresponse, _pool\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    880\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m MaxRetryError:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/urllib3/util/retry.py:592\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m new_retry\u001b[38;5;241m.\u001b[39mis_exhausted():\n\u001b[0;32m--> 592\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m MaxRetryError(_pool, url, error \u001b[38;5;129;01mor\u001b[39;00m ResponseError(cause))\n\u001b[1;32m    594\u001b[0m log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIncremented Retry for (url=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m): \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, url, new_retry)\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPSConnectionPool(host='trends.google.com', port=443): Max retries exceeded with url: /trends/api/widgetdata/comparedgeo?req=%7B%22geo%22%3A+%7B%22country%22%3A+%22US%22%7D%2C+%22comparisonItem%22%3A+%5B%7B%22time%22%3A+%222011-01-01+2011-12-31%22%2C+%22complexKeywordsRestriction%22%3A+%7B%22keyword%22%3A+%5B%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22car+stolen%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22find+stolen+car%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22report+police+stolen+car%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22insurance+car+stolen-dream-check%22%7D%5D%2C+%22operator%22%3A+%22OR%22%7D%7D%5D%2C+%22resolution%22%3A+%22DMA%22%2C+%22locale%22%3A+%22en-US%22%2C+%22requestOptions%22%3A+%7B%22property%22%3A+%22%22%2C+%22backend%22%3A+%22IZG%22%2C+%22category%22%3A+0%7D%2C+%22userConfig%22%3A+%7B%22userType%22%3A+%22USER_TYPE_SCRAPER%22%7D%2C+%22includeLowSearchVolumeGeos%22%3A+true%7D&token=APP6_UEAAAAAZd86w2fv-SZP28rOlNlhEpDIhsrwrJX-&tz=360 (Caused by ResponseError('too many 429 error responses'))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mRetryError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#Get 100 times weekly GT MVT data per city\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m times \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m100\u001b[39m):\n\u001b[0;32m----> 3\u001b[0m     df \u001b[38;5;241m=\u001b[39m gt_by_region_by_year(mvt_keyword, start_year, end_year, geo_level)\n\u001b[1;32m      4\u001b[0m     df\u001b[38;5;241m.\u001b[39mto_csv(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgeo_level\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstart_year\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mend_year\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_annually_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpd\u001b[38;5;241m.\u001b[39mTimestamp\u001b[38;5;241m.\u001b[39mnow()\u001b[38;5;241m.\u001b[39mstrftime(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mY\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mm\u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mH_\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mM\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgeo_level\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mstart_year\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mend_year\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_annually_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpd\u001b[38;5;241m.\u001b[39mTimestamp\u001b[38;5;241m.\u001b[39mnow()\u001b[38;5;241m.\u001b[39mstrftime(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mY\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mm\u001b[39m\u001b[38;5;132;01m%d\u001b[39;00m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mH_\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mM\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m file saved\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[6], line 6\u001b[0m, in \u001b[0;36mgt_by_region_by_year\u001b[0;34m(mvt_keyword, start_year, end_year, geo_level)\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m      5\u001b[0m     pytrend\u001b[38;5;241m.\u001b[39mbuild_payload(mvt_keyword, cat\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m, timeframe\u001b[38;5;241m=\u001b[39m\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m-01-01 \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m-12-31\u001b[39m\u001b[38;5;124m\"\u001b[39m, geo \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUS\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 6\u001b[0m     temp_df \u001b[38;5;241m=\u001b[39m pytrend\u001b[38;5;241m.\u001b[39minterest_by_region(resolution\u001b[38;5;241m=\u001b[39mgeo_level, inc_low_vol\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, inc_geo_code\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m      7\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mObtaining GT MVT from \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m-01-01 to \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m-12-31 in \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgeo_level\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ResponseError, Timeout, \u001b[38;5;167;01mConnectionAbortedError\u001b[39;00m, \u001b[38;5;167;01mConnectionError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m error:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pytrends/request.py:340\u001b[0m, in \u001b[0;36mTrendReq.interest_by_region\u001b[0;34m(self, resolution, inc_low_vol, inc_geo_code)\u001b[0m\n\u001b[1;32m    337\u001b[0m region_payload[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtz\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtz\n\u001b[1;32m    339\u001b[0m \u001b[38;5;66;03m# parse returned json\u001b[39;00m\n\u001b[0;32m--> 340\u001b[0m req_json \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_data(\n\u001b[1;32m    341\u001b[0m     url\u001b[38;5;241m=\u001b[39mTrendReq\u001b[38;5;241m.\u001b[39mINTEREST_BY_REGION_URL,\n\u001b[1;32m    342\u001b[0m     method\u001b[38;5;241m=\u001b[39mTrendReq\u001b[38;5;241m.\u001b[39mGET_METHOD,\n\u001b[1;32m    343\u001b[0m     trim_chars\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m5\u001b[39m,\n\u001b[1;32m    344\u001b[0m     params\u001b[38;5;241m=\u001b[39mregion_payload,\n\u001b[1;32m    345\u001b[0m )\n\u001b[1;32m    346\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(req_json[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdefault\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgeoMapData\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m    347\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (df\u001b[38;5;241m.\u001b[39mempty):\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pytrends/request.py:140\u001b[0m, in \u001b[0;36mTrendReq._get_data\u001b[0;34m(self, url, method, trim_chars, **kwargs)\u001b[0m\n\u001b[1;32m    136\u001b[0m     response \u001b[38;5;241m=\u001b[39m s\u001b[38;5;241m.\u001b[39mpost(url, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout,\n\u001b[1;32m    137\u001b[0m                       cookies\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcookies, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    138\u001b[0m                       \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrequests_args)  \u001b[38;5;66;03m# DO NOT USE retries or backoff_factor here\u001b[39;00m\n\u001b[1;32m    139\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 140\u001b[0m     response \u001b[38;5;241m=\u001b[39m s\u001b[38;5;241m.\u001b[39mget(url, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout, cookies\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcookies,\n\u001b[1;32m    141\u001b[0m                      \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrequests_args)  \u001b[38;5;66;03m# DO NOT USE retries or backoff_factor here\u001b[39;00m\n\u001b[1;32m    142\u001b[0m \u001b[38;5;66;03m# check if the response contains json and throw an exception otherwise\u001b[39;00m\n\u001b[1;32m    143\u001b[0m \u001b[38;5;66;03m# Google mostly sends 'application/json' in the Content-Type header,\u001b[39;00m\n\u001b[1;32m    144\u001b[0m \u001b[38;5;66;03m# but occasionally it sends 'application/javascript\u001b[39;00m\n\u001b[1;32m    145\u001b[0m \u001b[38;5;66;03m# and sometimes even 'text/javascript\u001b[39;00m\n\u001b[1;32m    146\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m200\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mapplication/json\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m \\\n\u001b[1;32m    147\u001b[0m         response\u001b[38;5;241m.\u001b[39mheaders[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mContent-Type\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;129;01mor\u001b[39;00m \\\n\u001b[1;32m    148\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mapplication/javascript\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m response\u001b[38;5;241m.\u001b[39mheaders[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mContent-Type\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;129;01mor\u001b[39;00m \\\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    151\u001b[0m     \u001b[38;5;66;03m# some responses start with garbage characters, like \")]}',\"\u001b[39;00m\n\u001b[1;32m    152\u001b[0m     \u001b[38;5;66;03m# these have to be cleaned before being passed to the json parser\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/requests/sessions.py:602\u001b[0m, in \u001b[0;36mSession.get\u001b[0;34m(self, url, **kwargs)\u001b[0m\n\u001b[1;32m    594\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request. Returns :class:`Response` object.\u001b[39;00m\n\u001b[1;32m    595\u001b[0m \n\u001b[1;32m    596\u001b[0m \u001b[38;5;124;03m:param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[1;32m    597\u001b[0m \u001b[38;5;124;03m:param \\*\\*kwargs: Optional arguments that ``request`` takes.\u001b[39;00m\n\u001b[1;32m    598\u001b[0m \u001b[38;5;124;03m:rtype: requests.Response\u001b[39;00m\n\u001b[1;32m    599\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    601\u001b[0m kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m--> 602\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrequest(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGET\u001b[39m\u001b[38;5;124m\"\u001b[39m, url, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msend(prep, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39msend_kwargs)\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m adapter\u001b[38;5;241m.\u001b[39msend(request, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/requests/adapters.py:510\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    507\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m ConnectTimeout(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    509\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e\u001b[38;5;241m.\u001b[39mreason, ResponseError):\n\u001b[0;32m--> 510\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m RetryError(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    512\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e\u001b[38;5;241m.\u001b[39mreason, _ProxyError):\n\u001b[1;32m    513\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ProxyError(e, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "\u001b[0;31mRetryError\u001b[0m: HTTPSConnectionPool(host='trends.google.com', port=443): Max retries exceeded with url: /trends/api/widgetdata/comparedgeo?req=%7B%22geo%22%3A+%7B%22country%22%3A+%22US%22%7D%2C+%22comparisonItem%22%3A+%5B%7B%22time%22%3A+%222011-01-01+2011-12-31%22%2C+%22complexKeywordsRestriction%22%3A+%7B%22keyword%22%3A+%5B%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22car+stolen%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22find+stolen+car%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22report+police+stolen+car%22%7D%2C+%7B%22type%22%3A+%22BROAD%22%2C+%22value%22%3A+%22insurance+car+stolen-dream-check%22%7D%5D%2C+%22operator%22%3A+%22OR%22%7D%7D%5D%2C+%22resolution%22%3A+%22DMA%22%2C+%22locale%22%3A+%22en-US%22%2C+%22requestOptions%22%3A+%7B%22property%22%3A+%22%22%2C+%22backend%22%3A+%22IZG%22%2C+%22category%22%3A+0%7D%2C+%22userConfig%22%3A+%7B%22userType%22%3A+%22USER_TYPE_SCRAPER%22%7D%2C+%22includeLowSearchVolumeGeos%22%3A+true%7D&token=APP6_UEAAAAAZd86w2fv-SZP28rOlNlhEpDIhsrwrJX-&tz=360 (Caused by ResponseError('too many 429 error responses'))"
     ]
    }
   ],
   "source": [
    "\n",
    "#Get 100 times weekly GT MVT data per city\n",
    "for times in range(100):\n",
    "    df = gt_by_region_by_year(mvt_keyword, start_year, end_year, geo_level)\n",
    "    df.to_csv(f\"{geo_level}_{start_year}_{end_year}_annually_{pd.Timestamp.now().strftime('%Y%m%d_%H_%M')}.csv\")\n",
    "    print(f\"{geo_level}_{start_year}_{end_year}_annually_{pd.Timestamp.now().strftime('%Y%m%d_%H_%M')} file saved\")\n",
    "    time.sleep(random.randint(10200, 10600))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "015e589e",
   "metadata": {},
   "outputs": [],
   "source": [
    "test.plot(figsize=(20, 12),  kind ='line')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edc81da1",
   "metadata": {},
   "outputs": [],
   "source": [
    "[print(i/25*23) for i in test2[\"car stolen+find stolen car+report police stolen car+insurance car stolen-dream-check\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffdf2371",
   "metadata": {},
   "outputs": [],
   "source": [
    "kw_list=['Facebook', 'Apple', 'Amazon', 'Netflix', 'Google']\n",
    "pytrend = TrendReq()\n",
    "#search interest per region\n",
    "#run model for keywords (can also be competitors)\n",
    "pytrend.build_payload(kw_list, timeframe='today 1-m')\n",
    "\n",
    "# Interest by Region\n",
    "regiondf = pytrend.interest_by_region()\n",
    "#looking at rows where all values are not equal to 0\n",
    "regiondf = regiondf[(regiondf != 0).all(1)]\n",
    "\n",
    "#drop all rows that have null values in all columns\n",
    "regiondf.dropna(how='all',axis=0, inplace=True)\n",
    "\n",
    "#visualise\n",
    "regiondf.plot(figsize=(20, 12), y=kw_list, kind ='bar')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecad1d3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "historicaldf = pytrend.get_historical_interest(['Facebook', 'Apple', 'Amazon', 'Netflix', 'Google'], \n",
    "                                               year_start=2020, month_start=1, day_start=1, hour_start=0, \n",
    "                                               year_end=2020, month_end=12, day_end=31, hour_end=0, \n",
    "                                               cat=0, geo='en-US', gprop='', sleep=0)\n",
    "\n",
    "#visualise\n",
    "#plot a timeseries chart\n",
    "historicaldf.plot(figsize=(20, 12))\n",
    "\n",
    "#plot seperate graphs, using theprovided keywords\n",
    "historicaldf.plot(subplots=True, figsize=(20, 12))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfd89f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gt_func(start_date_list, end_date_list, resolution, sample_size, path):\n",
    "#get monthly GT data\n",
    "    import inflect\n",
    "    #give ordinal names to numbers\n",
    "    p = inflect.engine()\n",
    "\n",
    "    for t in range(sample_size): #sample from GT 100 times\n",
    "        #time.sleep(random.randint(21600,21668)) # 6 hrs\n",
    "        print(f\"The {p.ordinal(t+1)} pull from GT\")\n",
    "        qt_list = [\"Q1\",\"Q2\",\"Q3\",\"Q4\"]\n",
    "        count = 0\n",
    "        for i, j in zip(start_date_list, end_date_list):\n",
    "            df = pd.DataFrame()\n",
    "            #Q1 - Q4\n",
    "            if count == 4:\n",
    "                count = 0\n",
    "            for k in kw_list:\n",
    "                #q = the name of the quarter\n",
    "                q = qt_list[count]\n",
    "                try:\n",
    "                    pytrend.build_payload(k, cat=0, timeframe=f'{i.strftime(\"%Y-%m-%d\")} {j.strftime(\"%Y-%m-%d\")}',  geo='US', gprop='')\n",
    "                    df = pd.concat([df, pytrend.interest_by_region(resolution=\"DMA\", inc_low_vol = False, inc_geo_code = False)], axis = 1)\n",
    "                    time.sleep(random.randint(120, 130))\n",
    "\n",
    "                except (Timeout, ResponseError) as e:\n",
    "                        print(f'!!{i.strftime(\"%Y-%m-%d\")} {j.strftime(\"%Y-%m-%d\")}!! ReadTimeoutError')\n",
    "                        print(f\"Whoops! ReadTimeoutError, the {p.ordinal(t+1)} pull aborted, now we take a rest\")\n",
    "                        print(dt.datetime.now().strftime(\"%Y/%m/%d_%H:%M:%S\"))\n",
    "                        time.sleep(random.randint(86668, 86888)) # 24 hrs\n",
    "                        print(f\"Rest enough, now we continue!\")\n",
    "\n",
    "                        #continue working\n",
    "                        pytrend.build_payload(k, cat=0, timeframe=f'{i.strftime(\"%Y-%m-%d\")} {j.strftime(\"%Y-%m-%d\")}',  geo='US', gprop='')\n",
    "                        df = pd.concat([df, pytrend.interest_by_region(resolution=\"DMA\", inc_low_vol=False, inc_geo_code=False)], axis = 1)\n",
    "                        time.sleep(random.randint(120, 130))\n",
    "\n",
    "            df.columns = [f\"MVT_{i.strftime('%Y')}_{q}\", \n",
    "                          f\"rape_{i.strftime('%Y')}_{q}\", \n",
    "                          f\"larceny_{i.strftime('%Y')}_{q}\", \n",
    "                          f\"burglary_{i.strftime('%Y')}_{q}\", \n",
    "                          f\"known_deliq_{i.strftime('%Y')}_{q}\", \n",
    "                          f\"low_social_cont_{i.strftime('%Y')}_{q}\"]\n",
    "            df.to_csv(f'{path}gt_crime_{i.strftime(\"%Y\")}_{q}_{dt.datetime.now().strftime(\"%Y%m%d_%H-%M-%S\")}.csv')\n",
    "            count += 1\n",
    "        print(f\"The {p.ordinal(t+1)} pull from GT Done!!\")\n",
    "        print(f'Time finished: {dt.datetime.now().strftime(\"%Y/%m/%d, %H:%M\")}')\n",
    "        time.sleep(random.randint(10800,10868)) # 3 hrs\n",
    "\n",
    "\n",
    "    print(f\"Congratulations, Well Done!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
